### PR TITLE
feat(GO1.21)!: support deletion of Tag Keys not in use in organizations in the clean up module

### PR DESCRIPTION
**Pull request by daniel-cit**
This pull request adds support for the deletion of Tag Keys not in use in organizations in the clean up module.
This PR also bumps the clean up Cloud Function runtime to golang `1.21`

### CODE REVIEW

File: modules/project_cleanup/function_source/main.go [hunk 1]
```diff
@@ -173,6 +193,36 @@ func getLabelsMapFromEnv(envVariableName string) map[string]string {
...
+	} else {
+		logger.Printf("Got Tag Keys list [%s] from [%s] env variable", tagKeys, envVariableName)
+	}
+	return tagKeys
+}
+
+func getCleanUpTagKeysOrTerminateExecution() bool {
+	cleanUpTagKeys, exists := os.LookupEnv(CleanUpTagKeys)
+	if !exists {
+		logger.Fatalf("Clean up Tag Keys flag not set [%s]. Specify correct value, Please.", CleanUpTagKeys)
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:217]: nit: Our typical error message style doesn't include "please".  Perhaps something instruction such as "Set the [%s] environment variable and try again."

If you really want to keep "please", I'd recommend something like: "Please set the required [%s] environment variable".

File: modules/project_cleanup/function_source/main.go [hunk 2]
```diff
@@ -173,6 +193,36 @@ func getLabelsMapFromEnv(envVariableName string) map[string]string {
...
+	}
+	result, err := strconv.ParseBool(cleanUpTagKeys)
+	if err != nil {
+		logger.Fatalf("Invalid Clean up Tag Keys flag [%s] value [%s]. Specify correct value, Please.", CleanUpTagKeys, cleanUpTagKeys)
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:221]: nit: similar "Please" here

File: modules/project_cleanup/function_source/main.go [hunk 3]
```diff
@@ -173,6 +193,36 @@ func getLabelsMapFromEnv(envVariableName string) map[string]string {
...
+	logger.Println("Try to get Tag Keys list")
+	if targetExcludedTagKeys == "" {
+		logger.Printf("No Tag Keys provided.")
+		return nil
+	}
+
+	var tagKeys []string
+	err := json.Unmarshal([]byte(targetExcludedTagKeys), &tagKeys)
+	if err != nil {
+		logger.Printf("Fail to get Tag Keys list from [%s] env variable, error [%s]", envVariableName, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:207]:
```suggestion
		logger.Printf("Failed to get Tag Keys list from [%s] env variable, error [%s]", envVariableName, err.Error())
```

File: modules/project_cleanup/function_source/main.go (hunk 1 above)
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:217]: Is this a cli flag or actually env variable?

```suggestion
		logger.Fatalf("Clean up Tag Keys env variable not set [%s]. Specify correct value, Please.", CleanUpTagKeys)
```

File: modules/project_cleanup/function_source/main.go (hunk 2 above)
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:221]: env variable?

```suggestion
		logger.Fatalf("Invalid Clean up Tag Keys env variable [%s] value [%s]. Specify correct value, Please.", CleanUpTagKeys, cleanUpTagKeys)
```

File: modules/project_cleanup/function_source/main.go [hunk 4]
```diff
@@ -182,6 +232,15 @@ func getCorrectFolderIdOrTerminateExecution() string {
 	return targetFolderIdString
 }
 
+func getCorrectOrganizationIdOrTerminateExecution() string {
+	targetOrganizationIdString := os.Getenv(TargetOrganizationId)
+	matched, err := regexp.MatchString(targetOrganizationRegexp, targetOrganizationIdString)
+	if err != nil || !matched {
+		logger.Fatalf("Invalid organization id [%s]. Specify correct value, Please.", targetOrganizationIdString)
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:239]: nit: same here

File: modules/project_cleanup/function_source/main.go [hunk 5]
```diff
@@ -210,6 +269,26 @@ func getFolderServiceOrTerminateExecution(client *http.Client) *cloudresourceman
 	return cloudResourceManagerService.Folders
 }
 
+func getTagKeysServiceOrTerminateExecution(client *http.Client) *cloudresourcemanager3.TagKeysService {
+	logger.Println("Try to get TagKeys Service")
+	cloudResourceManagerService, err := cloudresourcemanager3.New(client)
+	if err != nil {
+		logger.Fatalf("Fail to get TagKeys Service with error [%s], terminate execution", err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:276]:
```suggestion
		logger.Fatalf("Failed to get TagKeys Service with error [%s], terminate execution", err.Error())
```

File: modules/project_cleanup/function_source/main.go [hunk 6]
```diff
@@ -210,6 +269,26 @@ func getFolderServiceOrTerminateExecution(client *http.Client) *cloudresourceman
...
+	}
+	logger.Println("Got TagKeys Service")
+	return cloudResourceManagerService.TagKeys
+}
+
+func getTagValuesServiceOrTerminateExecution(client *http.Client) *cloudresourcemanager3.TagValuesService {
+	logger.Println("Try to get TagValues Service")
+	cloudResourceManagerService, err := cloudresourcemanager3.New(client)
+	if err != nil {
+		logger.Fatalf("Fail to get TagValues Service with error [%s], terminate execution", err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:286]:
```suggestion
		logger.Fatalf("Failed to get TagValues Service with error [%s], terminate execution", err.Error())
```

File: modules/project_cleanup/function_source/main.go [hunk 7]
```diff
@@ -247,6 +328,49 @@ func invoke(ctx context.Context) {
 		}
 	}
 
+	tagKeyAgeFilter := func(tagKey *cloudresourcemanager3.TagKey) bool {
+		tagKeyCreatedAt, err := time.Parse(time.RFC3339, tagKey.CreateTime)
+		if err != nil {
+			logger.Printf("Fail to parse CreateTime for tagKey [%s], skip it. Error [%s]", tagKey.Name, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:334]:
```suggestion
			logger.Printf("Failed to parse CreateTime for tagKey [%s], skip it. Error [%s]", tagKey.Name, err.Error())
```

Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:334]: suggestion: `skip` > `skipping`

File: modules/project_cleanup/function_source/main.go [hunk 8]
```diff
@@ -247,6 +328,49 @@ func invoke(ctx context.Context) {
...
+			return false
+		}
+		return tagKeyCreatedAt.Before(resourceCreationCutoff)
+	}
+
+	removeTagValues := func(tagKey string) {
+		logger.Printf("Try to remove Tag Values from TagKey [%s]", tagKey)
+		tagValuesList, err := tagValuesService.List().Parent(tagKey).Context(ctx).Do()
+		if err != nil {
+			logger.Printf("Fail to list Tag values from TagKey [%s], error [%s]", tagKey, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:344]:
```suggestion
			logger.Printf("Failed to list Tag values from TagKey [%s], error [%s]", tagKey, err.Error())
```

File: modules/project_cleanup/function_source/main.go [hunk 9]
```diff
@@ -247,6 +328,49 @@ func invoke(ctx context.Context) {
...
+			return
+		}
+		for _, tagValue := range tagValuesList.TagValues {
+			_, err := tagValuesService.Delete(tagValue.Name).Context(ctx).Do()
+			if err != nil {
+				logger.Printf("Fail to delete tagValue from TagKey [%s], error [%s]", tagKey, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:350]:
```suggestion
				logger.Printf("Failed to delete tagValue from TagKey [%s], error [%s]", tagKey, err.Error())
```

File: modules/project_cleanup/function_source/main.go [hunk 10]
```diff
@@ -247,6 +328,49 @@ func invoke(ctx context.Context) {
...
+			}
+		}
+	}
+
+	removeTagKeys := func(organization string) {
+		logger.Printf("Try to remove Tag Keys from organization [%s]", organization)
+		parent := fmt.Sprintf("organizations/%s", organization)
+		tagKeysList, err := tagKeyService.List().Parent(parent).Context(ctx).Do()
+		if err != nil {
+			logger.Printf("Fail to list Tag Keys from organization [%s], error [%s]", organization, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:360]:
```suggestion
			logger.Printf("Failed to list Tag Keys from organization [%s], error [%s]", organization, err.Error())
```

File: modules/project_cleanup/function_source/main.go [hunk 11]
```diff
@@ -247,6 +328,49 @@ func invoke(ctx context.Context) {
...
+			return
+		}
+		for _, tagKey := range tagKeysList.TagKeys {
+			if !checkIfTagKeyShortNameExcluded(tagKey.ShortName, excludedTagKeysList) && tagKeyAgeFilter(tagKey) {
+				removeTagValues(tagKey.Name)
+				_, err := tagKeyService.Delete(tagKey.Name).Context(ctx).Do()
+				if err != nil {
+					logger.Printf("Fail to delete tagKey from organization [%s], error [%s]", organization, err.Error())
```
Comment (apeabody) [inline modules/project_cleanup/function_source/main.go:368]:
```suggestion
					logger.Printf("Failed to delete tagKey from organization [%s], error [%s]", organization, err.Error())
```

File: modules/project_cleanup/main.tf [hunk 12]
```diff
@@ -52,14 +53,17 @@ module "scheduled_project_cleaner" {
   topic_name                     = var.topic_name
   function_available_memory_mb   = 128
   function_description           = "Clean up GCP projects older than ${var.max_project_age_in_hours} hours matching particular tags"
-  function_runtime               = "go113"
+  function_runtime               = "go121"
```
Comment (apeabody) [inline modules/project_cleanup/main.tf:56]: suggestion: Do we want to also bump https://github.com/terraform-google-modules/terraform-google-scheduled-function/blob/master/modules/project_cleanup/function_source/go.mod#L3 to go `1.21`?  Might also need a `mod tidy`.

File: modules/project_cleanup/variables.tf [hunk 13]
```diff
@@ -77,6 +77,18 @@ variable "target_included_labels" {
...
 
+variable "clean_up_org_level_tag_keys" {
+  type        = bool
+  description = "Clean up organization level Tag Keys."
+  default     = false
+}
+
+variable "target_excluded_tagkeys" {
+  type        = list(string)
+  description = "List of organization Tag Keys short names that won't be deleted."
```
Comment (apeabody) [inline modules/project_cleanup/variables.tf:88]: As "names" is plural, I'm not sure if we also need the "s" on "Key"?

```suggestion
  description = "List of organization Tag Key short names that won't be deleted."
```

**Review by apeabody — CHANGES_REQUESTED** [review]
Thanks @daniel-cit!  Made a few error message suggestions.

File: modules/project_cleanup/function_source/main.go (hunk 3 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go:207]: @apeabody we have another 18 cases of previously existing messages with `Fail to`. 
Should we update them all for consistency?

File: modules/project_cleanup/function_source/main.go (hunk 1 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: @apeabody we have another 2 cases of previously existing messages with "Please".
Should we update them all for consistency?

File: modules/project_cleanup/function_source/main.go (hunk 7 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: @apeabody we have another 2 cases of previously existing messages with "skip it".
Should we update them all for consistency?

File: modules/project_cleanup/variables.tf (hunk 13 above)
Comment (daniel-cit) [reply modules/project_cleanup/variables.tf]: fixed

File: modules/project_cleanup/main.tf (hunk 12 above)
Comment (daniel-cit) [reply modules/project_cleanup/main.tf:56]: done

File: modules/project_cleanup/function_source/main.go (hunk 1 above)
Comment (apeabody) [reply modules/project_cleanup/function_source/main.go]: Hi @daniel-cit - I would suggest doing so.

File: modules/project_cleanup/function_source/main.go (hunk 3 above)
Comment (apeabody) [reply modules/project_cleanup/function_source/main.go:207]: Hi @daniel-cit - If it's not too much trouble, I would do so. :)

File: modules/project_cleanup/main.tf (hunk 12 above)
Comment (apeabody) [reply modules/project_cleanup/main.tf:56]: Marked as a breaking change given the change in function runtime.  That said it's going into what is already a major release anyway.

File: modules/project_cleanup/function_source/main.go (hunk 1 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 2 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 3 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 9 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 10 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 11 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 4 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 5 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 6 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 7 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

File: modules/project_cleanup/function_source/main.go (hunk 8 above)
Comment (daniel-cit) [reply modules/project_cleanup/function_source/main.go]: fixed

**Review by apeabody — APPROVED** [review]
Thanks @daniel-cit - LGTM!

**Pull request merged**